modules/widgets/application-title-bar.nix

```diff
@@ -0,0 +1,446 @@
+  inherit (widgets.lib) mkBoolOption mkEnumOption;
+
+  convertHorizontalAlignment = horizontalAlignment: let
+    mappings = {
+      "left" = "1";
+      "right" = "2";
+      "center" = "4";
+      "justify" = "8";
+    };
+  in
+    mappings.${horizontalAlignment} or (throw "Invalid enum value: ${horizontalAlignment}");
+
+  convertVerticalAlignment = verticalAlignment: let
+    mappings = {
+      "top" = "1";
+      "center" = "128";
+      "bottom" = "64";
+      "baseline" = "256";
+    };
+  in
+    mappings.${verticalAlignment} or (throw "Invalid enum value: ${verticalAlignment}");
+
+  fontType = types.submodule {
+    options = {
+      bold = mkBoolOption "Enable bold text.";
+      fit = mkEnumOption [ "FixedSize" "HorizontalFit" "VerticalFit" "Fit" ] // {
+        example = "FixedSize";
+        description = "The mode of the size of the font.";
+      };
+      size = mkOption {
+        type = types.ints.positive;
+        default = 10;
+        description = "The size of the font.";
+        apply = builtins.toString;
+      };
+    };
+  };
+
+  marginType = types.submodule {
+    options = {
+      left = mkOption {
+        type = types.ints.unsigned;
+        default = 10;
+        description = "The left margin.";
+        apply = builtins.toString;
+      };
+      right = mkOption {
+        type = types.ints.unsigned;
+        default = 10;
+        description = "The right margin.";
+        apply = builtins.toString;
+      };
+      top = mkOption {
+        type = types.ints.unsigned;
+        default = 0;
+        description = "The top margin.";
+        apply = builtins.toString;
+      };
+      bottom = mkOption {
+        type = types.ints.unsigned;
+        default = 0;
+        description = "The bottom margin.";
+        apply = builtins.toString;
+      };
+    };
+  };
+in
+{
+  applicationTitleBar = {
+    description = "KDE plasmoid with window title and buttons";
+
+    opts = {
+      layout = {
+        widgetMargins = mkOption {
+          type = types.nullOr types.ints.unsigned;
+          default = null;
+          description = "The margins around the widget.";
+          apply = builtins.toString;
+        };
+        spacingBetweenElements = mkOption {
+          type = types.nullOr types.ints.unsigned;
+          default = null;
+          description = "The spacing between elements.";
+          apply = builtins.toString;
+        };
+        horizontalAlignment = mkOption {
+          type = types.nullOr (types.enum [ "left" "right" "center" "justify" ]);
+          default = null;
+          description = "The horizontal alignment of the widget.";
+          apply = convertHorizontalAlignment;
+        };
+        verticalAlignment = mkOption {
+          type = types.nullOr (types.enum [ "top" "center" "bottom" "baseline" ]);
+          default = null;
+          description = "The vertical alignment of the widget.";
+          apply = convertVerticalAlignment;
+        };
```

Comment:
These allow for `null` in the type (in fact it defaults to `null`), but the derivation crashes when these options are `null`. Either allow for `null` values in `convertVericalAlignment`/`convertHorizontalAlignment` or don't allow them to take the value `null` at all (and set some default values)

Reply:
Fixed. I didn't notice it when I was writing it.